style_guides/pl_sql_general_guidance.md

```diff
@@ -0,0 +1,175 @@
+# General PL/SQL Writing Guidance
+
+This page contains some general guidance on how to approach writing pl/sql, but it is not intended to be an exhaustive list of good practice.
+
+
+* [Column Aliases](#column-aliases)
+* [Table Joins](#table-joins)
+* [Transaction Control - COMMIT and ROLLBACK](#transaction-control-commit-and-rollback)
+* [Table Sequences](#table-sequences)
+* [Single Responsibility Principle](single-responsibility-principle)
+* [Conditional Statements](#conditional-statements)
+
+
+## Column Aliases
+
+Always rename aggregates and function-wrapped columns:
+
+```sql
+SELECT
+    name,
+    SUM(amount) AS sum_amount
+FROM ...
+```
+
+Always rename all columns when selecting with table aliases:
+
+```sql
+SELECT
+    projects.name AS project_name,
+    COUNT(backings.id) AS backings_count
+FROM 
+    ksr.backings AS backings
+    INNER JOIN ksr.projects AS projects ON ...
+```
+
+Always use `AS` to rename columns:
+
+__GOOD__:
+
+```sql
+SELECT
+    projects.name AS project_name,
+    COUNT(backings.id) AS backings_count
+...
+```
+
+__BAD__:
+
+```sql
+SELECT
+    projects.name project_name,
+    COUNT(backings.id) backings_count
+...
+```
+
+Order (`ASC`, `DESC`) should always be explicit. All window functions should be aliased.
+
+---
+
+## Table Joins
+
+__USE ANSI JOIN SYNTAX!__
+
+Explicitly use `INNER JOIN` not just `JOIN`, making multiple lines of `INNER JOIN`s easier to scan:
+
+__GOOD__:
+
+```sql
+SELECT
+    projects.name      AS project_name,
+    COUNT(backings.id) AS backings_count
+FROM 
+    ksr.projects AS projects
+    INNER JOIN ksr.backings AS backings
+    ON ...
+    LEFT OUTER JOIN ksr.backer_rewards AS backer_rewards 
+    ON ...
+    LEFT OUTER JOIN ...
+```
+
+__BAD__:
+
+```sql
+SELECT
+    projects.name      AS project_name,
+    COUNT(backings.id) AS backings_count
+FROM 
+    ksr.projects AS projects
+    JOIN ksr.backings AS backings 
+    ON ...
+    LEFT OUTER JOIN ksr.backer_rewards AS backer_rewards
+    ON ...
+    LEFT OUTER JOIN ...
+```
+
+---
+
+## Transaction Control COMMIT and ROLLBACK 
+
+A function or procedure should take a parameter to specify this behavior. This means it can be controlled by the calling code and makes it more flexible.

```

Comment:
Do we not need to consider all the other options to what if we want to do nothing. We are very limited in control if we are only able to rollback or commit. If we are enforcing having this in code and we are using things like automated testing tools then we need to be able to call something and do nothing  ?

Reply:
I also created a package in the past which could be called instead of including the large block of transaction code at the bottom, but i think we didn't want this in everything but it would reduce the amount of repeated control logic code in each package.

Reply:
I wouldn't pass in a parameter to control COMMIT and ROLLBCK but let the caller decide that.  The caller could be an Anonymous block or an entry point code.

Reply:
After conversation in teams I am removing content on this section and we will have a separate conversation focusing on this.